[plsql] ParseException while parsing EXECUTE IMMEDIATE 'drop database link ' || linkname;

It seems like PMD 6.31.0 does not like

    EXECUTE IMMEDIATE 'drop database link ' || linkname;

or more generally the case where the statement to be executed with EXECUTE IMMEDIATE is a concatenation:

Caused by: net.sourceforge.pmd.lang.plsql.ast.ParseException: Encountered " "||" "|| "" at line ..., column ....
Was expecting one of:
    ";" ...
    "INTO" ...
    "RETURN" ...
    "RETURNING" ...
    "USING" ...

Execute Immediate Statement Spec:
https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/EXECUTE-IMMEDIATE-statement.html#GUID-C3245A95-B85B-4280-A01F-12307B108DC8

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -2686,7 +2686,7 @@ ASTFetchStatement FetchStatement()  :
 ASTEmbeddedSqlStatement EmbeddedSqlStatement()  :
 {}
 {
-   <EXECUTE> <IMMEDIATE> (StringLiteral() | Expression())
+   <EXECUTE> <IMMEDIATE> Expression()
 	[ <INTO> Name() ("," Name())* ]
 	[ <USING> [ <IN> [ <OUT> ] | <OUT> ]  Expression() ("," [ <IN> [ <OUT> ] | <OUT> ] Expression())* ]
 	[ ( <RETURN> | <RETURNING> )  <INTO> Expression() ("," Expression())*] ";"

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -2686,7 +2686,7 @@ ASTFetchStatement FetchStatement()  :
 ASTEmbeddedSqlStatement EmbeddedSqlStatement()  :
 {}
 {
-   <EXECUTE> <IMMEDIATE> Expression()
+   <EXECUTE> <IMMEDIATE> StringExpression()
 	[ <INTO> Name() ("," Name())* ]
 	[ <USING> [ <IN> [ <OUT> ] | <OUT> ]  Expression() ("," [ <IN> [ <OUT> ] | <OUT> ] Expression())* ]
 	[ ( <RETURN> | <RETURNING> )  <INTO> Expression() ("," Expression())*] ";"

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/PLSQLParserTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/PLSQLParserTest.java
@@ -73,4 +73,9 @@ public class PLSQLParserTest extends AbstractPLSQLParserTst {
     public void testCaseIssue1454() {
         plsql.parseResource("CaseIssue1454.pls");
     }
+
+    @Test
+    public void testCaseIssue3106() {
+        plsql.parse("CREATE OR REPLACE PROCEDURE bar\nIS\n  v_link varchar2(10) := 'xxx';\nBEGIN\n  EXECUTE IMMEDIATE 'drop database link ' || v_link;\nEND bar;");
+    }
 }

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/PLSQLParserTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/PLSQLParserTest.java
@@ -75,7 +75,7 @@ public class PLSQLParserTest extends AbstractPLSQLParserTst {
     }
 
     @Test
-    public void testCaseIssue3106() {
-        plsql.parse("CREATE OR REPLACE PROCEDURE bar\nIS\n  v_link varchar2(10) := 'xxx';\nBEGIN\n  EXECUTE IMMEDIATE 'drop database link ' || v_link;\nEND bar;");
+    public void testExecuteImmediateIssue3106() {
+        plsql.parseResource("ExecuteImmediateIssue3106.pls");
     }
 }

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediateIssue3106.pls
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/ExecuteImmediateIssue3106.pls
@@ -1,0 +1,11 @@
+--
+-- BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+--
+
+CREATE OR REPLACE PROCEDURE bar
+IS
+    v_link varchar2(10) := 'xxx';
+BEGIN
+    -- EXECUTE IMMEDIATE with an expression instead of just a literal or variable.
+    EXECUTE IMMEDIATE 'drop database link ' || v_link;
+END bar;

